debian/man/iqrfgd2-tokenctl.1
.TH IQRFGD2-TOKENCTL 1 2026-01-28 "IQRF Gateway Daemon Token Control" "IQRF GW Manual"
.SH NAME
iqrfgd2-tokenctl \- manage IQRF Gateway Daemon API tokens
.SH DESCRIPTION
The \fBiqrfgd2-tokenctl\fP command allows for managing and listing of API tokens used for WebSocket authentication.
.SH SYNOPSIS
\fBiqrfgd2-tokenctl\fP [\fBcommand\fP] [\fIoptions\fP...]
.SH COMMANDS
.nf
.TP
.BR help\fP
Shows the help message and usage.
.TP
.BR list\fP " [\-p, \-\-path <PATH>]"
Outputs a list of API tokens. The output does not contain hashed secret or salt.
.TP
.BR get\fP " [\-i, \-\-id <TOKEN ID>] [\-p, \-\-path <PATH>]"
Outputs a token specified by ID. The output does not contain hashed secret or salt.
.TP
.BR create\fP " [\-o, \-\-owner <OWNER>] [\-e, \-\-expiration <EXPIRATION>] [\-s, \-\-service] [\-p, \-\-path <PATH>]"
Creates a new API token with owner, expiration and optionally permissions for service mode.
If service mode permissions are not desired, the option can simply be omitted.
.TP
.BR rotate\fP " [\-i, \-\-id <TOKEN ID>] [\-p, \-\-path <PATH>]"
Performs a token rotation. Rotating a token means its revocation, and a new token is created to be used instead.
If a token is already expired or revoked, rotation is not performed.
.TP
.BR revoke\fP " [\-i, \-\-id <TOKEN ID>] [\-p, \-\-path <PATH>]"
Revokes a token.
.SH OPTIONS
.nf
.TP
.BR "General options"
.TQ
.BR \-h ", " \-\-help
Shows the help message and usage of a command.
.TP
.BR \-j ", " \-\-json
Format output to stdout as JSON if applicable to the command.
.TP
.BR \-p ", " \-\-path
Specifies path to authentication database file.
.TP
.BR "Command options"
.TQ
.BR \-i ", " \-\-id
Token ID. Integer value coresponding to rowid of record in database.
.TP
.BR \-o ", " \-\-owner
Owner of the token. A string value representing intended user, client or application.
.TP
.BR \-e ", " \-\-expiration
Token expiration. Expiration can either be defined as unix epoch in seconds, or as relative time from the moment of invocation.
The format of relative expiration is defined by time unit and count. The format is \fBcount\fP\funit\fP.
Available units are:
.RS
.IP \[bu] 2
d - day
.IP \[bu] 2
w \- week
.IP \[bu] 2
m \- month
.IP \[bu] 2
y \- year
.TP
.RE
Example relative expiration: 30d \- token expires in 30 days.
.SH EXAMPLES
.TP
.BR "iqrfgd2-tokenctl list -j -p database.db"
Outputs a list of tokens in database serialized into JSON.
.TP
.BR "iqrfgd2-tokenctl -o user -e 30d -s -p database.db"
Creates a token with owner set as "user", that expires in 30 days and has service mode permissions.
.TP
.BR "iqrfgd2-tokenctl -i 5 rotate -p database.db"
Rotates a token with ID 5.
.SH LICENCE
This program is licensed under Apache License version 2.0.
.SH COPYRIGHT
.nf
Copyright 2015\-2026 IQRF Tech s.r.o.
Copyright 2019\-2026 MICRORISC s.r.o.
.fi


debian/changelog
iqrf-gateway-daemon (2.6.7~alpha~1.gbp6059b0) UNRELEASED; urgency=medium

  ** SNAPSHOT build @6059b0317d3d25af8000997d6913e5d8aa0c9697 **

  * UNRELEASED

 -- Karel Hanák <karel.hanak@iqrf.org>  Thu, 10 Jul 2025 15:21:25 +0200

iqrf-gateway-daemon (2.6.6) stable; urgency=medium

  [ Karel Hanák ]
  * Prepare for version 2.6.6-alpha

  [ Roman Ondráček ]
  * GitLab CI: build packages for Debian 13 trixie

  [ Karel Hanák ]
  * JsCache: replaced filesystem rename with copy and remove for cross-device data manipulation

 -- Karel Hanák <karel.hanak@iqrf.org>  Thu, 10 Jul 2025 15:19:07 +0200

iqrf-gateway-daemon (2.6.5) stable; urgency=medium

  [ Karel Hanák ]
  * Prepare for version 2.6.5-alpha
  * Scheduler: sync directory metadata when task file is removed

 -- Karel Hanák <karel.hanak@iqrf.org>  Tue, 08 Jul 2025 09:42:41 +0200

iqrf-gateway-daemon (2.6.4) stable; urgency=medium

  [ Karel Hanák ]
  * Prepare for version 2.6.4-alpha
  * api: allow unrequired response in light responses
  * IqrfInfo: always attempt to load old and new dali/light driver

 -- Karel Hanák <karel.hanak@iqrf.org>  Fri, 02 May 2025 14:46:14 +0200

iqrf-gateway-daemon (2.6.3) stable; urgency=medium

  * Prepare for version 2.6.3-alpha
  * api: fixed infoDaemon request and response schemas

 -- Karel Hanák <karel.hanak@iqrf.org>  Wed, 09 Apr 2025 11:18:41 +0200

iqrf-gateway-daemon (2.6.2) stable; urgency=medium

  * Prepare for version 2.6.2-alpha
  * OtaUpload: wait for OS batch writes to finish

 -- Karel Hanák <karel.hanak@iqrf.org>  Thu, 16 Jan 2025 15:21:18 +0100

iqrf-gateway-daemon (2.6.1) stable; urgency=medium

  [ Karel Hanák ]
  * Prepare for version 2.6.1-alpha
  * PingService: fixed implementation for HWPID filtering
  * OtaUpload: fixed incorrect code block length
  * updated license headers

 -- Karel Hanák <karel.hanak@iqrf.org>  Sun, 05 Jan 2025 11:51:30 +0100

iqrf-gateway-daemon (2.6.0) stable; urgency=medium

  [ Roman Ondráček ]
  * Update cpr, date and valijson to the latest versions
  * CMake: use system curl libraries for cpr
  * Revert "CMake: use system curl libraries for cpr"

 -- Karel Hanák <karel.hanak@iqrf.org>  Tue, 05 Nov 2024 06:13:32 -0500

iqrf-gateway-daemon (2.6.0~rc1) testing; urgency=medium

  * fixed iqmesh service responses not being sent if validation failed
  * iqmeshNetwork_RemoveBond: fixed network node count property in response api schema
  * added pnum and pcmd resolution in case request drivers are not available
  * removed old iqrfLight api schemas, examples and wrapper functions
  * removed old iqrfInfo dali schemas and examples, updated light schemas and examples, fixed metadata in schemas
  * debian package: changed cache content and schemas install dirs
  * debian package: install serverCheck file to cache dir
  * Autonetwork: reset exclusive access in case of unexpected failure

 -- Karel Hanák <karel.hanak@iqrf.org>  Mon, 21 Oct 2024 12:02:39 +0200

iqrf-gateway-daemon (2.6.0~beta2) stable; urgency=medium

  [ Karel Hanák ]
  * set logging level of deployment configurations to warning

 -- Karel Hanák <karel.hanak@iqrf.org>  Tue, 17 Sep 2024 10:24:17 +0200

iqrf-gateway-daemon (2.6.0~beta1) stable; urgency=medium

  [ Karel Hanák ]
  * Prepare for version 2.6.0-alpha
  * MqttMessaging: set cleanSession according to qos value
  * JsonDpaApiIqrfStdExt: fixed an error causing crashes when standard frc request is sent in to daemon in service mode
  * IdeCounterpart: include GW_ADDR in asynchronous TR packets
  * changed api testing image to latest

  [ Roman Ondráček ]
  * CMake: rework version handling

  [ Karel Hanák ]
  * AutonetworkService: always report new nodes
  * MonitorService: added notification worker invocation message: IdeCounterpart: mode set callbacks
  * websocket: set deployment configuration to accept connections from localhost only
  * api: added ntfDaemon_InvokeMonitor examples
  * MonitorService: unregister mode change callback before removing interface
  * updated license headers

  [ Roman Ondráček ]
  * Debian packaging: use xz compression type for package packing
  * Debian packaging: add gbp config

  [ Karel Hanák ]
  * JsonDpaApiIqrfStandard: pass error message to response when driver request error occurs
  * JsRenderDuktape: use default HWPID context as fallback in case of missing driver peripheral namespaces for address 0
  * IqrfInfo: always load coordinator device drivers
  * JsCache: cache updater, cache download and initialization rework, validation, support for quantities and metadata
  * JsonDpaApiIqrfStdExt: include DALI command in FRC response
  * JsonDpaApiIqrfStdExt: use nlohmann to process frc responses, fixed selected nodes
  * JsCache: parse metadata as array of objects, allow extra data in metadata
  * clibdpa: updated submodule reference
  * api: updated DALI FRC examples
  * api: updated Sensor FRC examples
  * API: updated iqmeshNetwork_RemoveBondOnlyInC examples
  * API: iqmeshNetwork_RemoveBondOnlyInC deviceAddr made optional
  * JsCache: added support for power supply metadata, updated jsonschemas and metadata structure
  * metadata: fixed version condition in jsonschema
  * JsCache: support for floating point values of product voltage properties

  [ Roman Ondráček ]
  * GitLab CI: build package for Ubuntu 24.04 Noble Numbat
  * CMake: fix build on Ubuntu 24.04
  * Update iqrf-header-parser and fix build on Ubuntu 24.04

  [ Karel Hanák ]
  * Added support for new IQRF Light standard

  [ Roman Ondráček ]
  * Docker: update deocker images

  [ Karel Hanák ]
  * include device mid and hwpid in sensor/dali/light frc responses with extended format
  * iqrfLight_SetLai: accept voltage as number in volts

  [ Roman Ondráček ]
  * .mailmap: update mappings
  * StringUtils: fix name of startsWith method, add missing document comments
  * MqttMessaging: autodetect TLS usage from MQTT broker URI protocol

 -- Roman Ondráček <mail@romanondracek.cz>  Tue, 20 Aug 2024 02:58:47 +0200


completions/iqrfgd2-tokenctl.bash
# iqrfgd2-tokenctl(1) completion                                      -*- shell-script -*-

_iqrfgd2-tokenctl() {
  local curr prev
  curr="${COMP_WORDS[COMP_CWORD]}"

  # available commands
  local commands="
    help
    list
    get
    create
    revoke
    rotate
  "
  # command options
  local shared_opts="-h --help -p --path"
  local list_opts="$shared_opts"
  local get_rotate_opts="$shared_opts -i --id -j --json"
  local create_opts="$shared_opts -o --owner -e --expiration -j --json"
  local revoke_opts="$shared_opts -i --id"

  # no command specified yet
  if [[ $COMP_CWORD == 1 ]]; then
    COMPREPLY=( $(compgen -W "$commands" -- "$curr") )
    return 0
  fi

  # command received
  cmd="${COMP_WORDS[1]}"
  curr="${COMP_WORDS[COMP_CWORD]}"
  prev="${COMP_WORDS[COMP_CWORD-1]}"

  # help found, stop completing
  for token in "${COMP_WORDS[@]}"; do
    case "$token" in
      -h|--help)
        COMPREPLY=()
        return
        ;;
    esac
  done

  # if specifying database file, look for db files
  case "$prev" in
    -p|--path)
      local files=( $(compgen -f -- "$curr") )
      COMPREPLY=()

      for file in "${files[@]}"; do
        if [[ -d "$file" ]]; then
          compopt -o nospace 2>/dev/null
          COMPREPLY+=( "$file/" )
          continue
        fi

        if [[ "$file" == *.db ]]; then
          COMPREPLY+=( "$file" )
        fi
      done
      return 0
      ;;
  esac

  case "$cmd" in
    help)
      COMPREPLY=()
      return
      ;;
    list)
      COMPREPLY=( $(compgen -W "$list_opts" -- "$curr") )
      return
      ;;
    get|rotate)
      COMPREPLY=( $(compgen -W "$get_rotate_opts" -- "$curr") )
      return
      ;;
    create)
      COMPREPLY=( $(compgen -W "$create_opts" -- "$curr") )
      return
      ;;
    revoke)
      COMPREPLY=( $(compgen -W "$revoke_opts" -- "$curr") )
      return
      ;;
  esac
}

complete -F _iqrfgd2-tokenctl iqrfgd2-tokenctl

# ex: filetype=sh
